### ISSUE TITLE
Security mode throughput is periodic in security.cert.batch.size

### ISSUE THREAD

**Issue by bcbrock** [hyperledger-archives/fabric#996]
I was asking questions on Slack about the poor system-level throughput with security enabled. @muralisrini suggested setting security.cert.batch.size to 1000 and ignoring the throughput of the initial transactions that "bring down the tcerts". In doing these experiments on a Docker-compose network consisting of a single validating peer running NOOPS + OBC-CA, I observed the behavior shown in the following graph. 

Transactions are carefully grouped by the driver into groups of 200 (including invoke, deploy and query). Groups that contain the 1st, 1001st, 2001st etc. transactions have the lowest throughput. Surprisingly, the throughput increases dramatically as the transaction indices increase within a group of 1000.  I am not familiar with the algorithms involved but this looks like some kind of O(N) algorithm is being run based on the transaction index. Looking at `top` it also appears that the utilization decreases in the obc-peer process as the transaction index modulo 1000 increases. 

Unfortunately the throughput is only better with higher indexes, and is still never very good. I've opened this issue in the hope that it might prompt someone to figure out and fix part of the performance shortfall.

Note: I'm still using recent OBC code; I haven't converted my drivers to HyperLedger yet.

![txrate](https://cloud.githubusercontent.com/assets/11968347/14300478/3b343bd0-fb56-11e5-9a7a-1e4d8a7a7913.png)

**Comment by bcbrock** [comment]
More details: There is a single simple chaincode, similar to chaincode_example02 (but slightly different).  The first block of 200 TX is 1 deploy, 198 invokes, 1 query.  All but the first block of 200 TX consist of 199 invokes plus 1 query. [The queries are correctness checks at the end]. Throughput is measured for each complete block of  200 TX - each block of 200 TX is considered as a separate test against a network that remains deployed between tests.

IBMers can recreate by
1. git clone https://github.ibm.com/bcbrock/busywork
2. Install as directed in the README.md
3. cd ... bcbrock/busywork/counters
4. make issue996

**Comment by adecaro** [comment]
@bcbrock, which golang version are you using?

**Comment by bcbrock** [comment]
@adecaro @muralisrini I am using golang version 1.6rc2 (built from source before Go 1.6 was officially released; I should upgrade this) on a 4-socket X86 server. This test is essentially the only work running on the system.

The transactions are generated by a single client that issues transactions as fast as they are accepted by the peer. Once all transactions have been issued the driver waits 10 seconds before issuing the correctness query. The throughput number does not include the correctness query or the 10-second wait.

**Comment by bcbrock** [comment]
Here are other ways of observing the effects shown above. This chart shows transaction throughput for this type of test w/o security, while varying with the number of simultaneous clients. Throughput tops out above 500 TX/second with 4 clients before flattening out. (This is 1 peer running NOOPS).

![chart1](https://cloud.githubusercontent.com/assets/11968347/14337644/8413dfa6-fc35-11e5-9398-2fa138391cb6.png)

This next chart is typical of why people are concerned about the throughput of secure mode. With the default setting of `security.tcert.batch.size = 200`, throughput appears to be very poor, less than 15 TX/second even with large numbers of clients.

![chart2](https://cloud.githubusercontent.com/assets/11968347/14337649/8b54463e-fc35-11e5-81cf-bcf25f5bb4d7.png)

However, this shortfall is due to what appears to be extreme inefficiencies in the generation of the certificates. In this chart, each run first generates a few more certificates than will be needed for the run. The throughput plotted here excludes the first 3 transactions (a deploy, an invoke, and a query). Once past certificate generation, throughput is actually pretty good as shown below.

![chart3](https://cloud.githubusercontent.com/assets/11968347/14337663/91aea196-fc35-11e5-9031-f2819c5b75b9.png)

Digging deeper, during the certificate generation process, which may take several minutes when thousands are being generated, `top` shows that the peer and the obc-ca are virtually idle, each typically using no more than 5 or 6 percent of an X86 thread. The I/O write rate is approximately 1.5 MB/sec during this period - almost noise for this machine that can move in excess of 150 MB/sec. of I/O. So we need to figure out what's going on here. Another idea - is it possible to restructure the code to generate and use certificates concurrently?

Back in the high-throughput area, the logs indicate that the throughput of an individual client is being throttled by `crypto/client.go InitClient()` which takes between 50 - 150 milliseconds to execute. One thing I noticed is that the client objects are continuously being recreated for the same user, which at first glance seems very wasteful. The actual sign/verify of the transactions takes virtually no time, I assume because the transactions are very small here. A profile of this high throughput area shows nothing involving crypto or hashing. I've attached a PDF of the Go profile of this area below.

[core.pdf](https://github.com/hyperledger/fabric/files/207576/core.pdf)

**Comment by muralisrini** [comment]
Thanks, @bcbrock . These charts with concurrent clients do shed more light.

With regards to no security with noops. Flattening at 500 is suggestive - that's the default "block.size" in consensus/noops/config.yaml. Wonder what we'd see if we raise it to 1000.

_top shows that the peer and the obc-ca are virtually idle, each typically using no more than 5 or 6 percent of an X86 thread. The I/O write rate is approximately 1.5 MB/sec during this period - almost noise for this machine that can move in excess of 150 MB/sec. of I/O. So we need to figure out what's going on here. Another idea - is it possible to restructure the code to generate and use certificates concurrently?_ 

Does suggest serial generation....

From a previous comment  
_Once all transactions have been issued the driver waits 10 seconds before issuing the correctness query. The throughput number does not include the correctness query or the 10-second wait._

Is that method followed for the multiple client measurements as well ? Polling query request at a small interval (say 1 sec) after the clients have completed submitting transactions might yield a tighter result.

**Comment by bcbrock** [comment]
@muralisrini - The "500" seems to be a coincidence. Going to blocks of 1000 doesn't improve throughput - it reduces it slightly.

In the measurements reported just above the timing protocol was changed slightly. Each client waits until it has confirmed that all of its transactions are registered in the blockchain before exiting. The confirmation adds a small latency (haven't tried to measure it though.)

@adecaro - It was suggested to me that I ask for your thoughts. Do you have a theory on what's causing the sawtooth pattern, or why tCert generation is not CPU bound? Why does InitClient() take so long, and why is it necessary to recreate the client object on every transaction?

For the record, someone suggested that tCert generation might be throttled by a lack of "entropy" that causes the random number generator to block. This does not seem to be an issue so far, it looks like we're getting a random number in no more than 10 microseconds on average on this system.

**Comment by adecaro** [comment]
@bcbrock, sorry for the late reply. I think the sawtooth pattern is very compatible with the way the client pre-load TCerts. Currently, there is a queue that is filled at very beginning (InitClient and the reason why it takes time) and once the queue is empty it is refilled again (this is blocking process)
Now, actually, there is a more agressive way to pre-load TCerts that is enabled by setting the propoerty security.multithreading.enabled to true in core.yaml. If this is enabled, then a thread is activated to load TCerts in background. Please, may you add that property and plot the results?

Regarding InitClient, actually it is enough to call it only once and cache the object the function returns. 

For TCert generation, @diegomasini is the right person for this. He is in charge of the membership service.

**Comment by bcbrock** [comment]
@adecaro - I tried security.multithreading.enabled=true  Unfortunately performance is much worse, going from a max of 5 TX/second to a constant 1 TX every 5 seconds in my testing.

If I understand correctly your comment about InitClient(), do you plan to fix this?

@diegomasini - Can you explain why we see such low utilization of the peer and member services processes when the TCerts are being generated? If this process was gated by crypto performance I would expect to see high CPU utilization.

**Comment by adecaro** [comment]
@bcbrock, @muralisrini, I found the problem. It is related to the way the crypto layer is used. We already discussed this in the past. 

Essentially, every time a transaction arrives, crypto.InitClient(...) is invoked and when the transaction has been processed, crypto.CloseClient(...) is invoked. Actually, it is not necessary to invoke crypto.CloseClient every time because it wastes a lot of time and makes a subsequent call to crypto.InitClient even slower. 

So, what I did was to comment out all the calls to crypto.CloseClient(...) which are in devops.go and rest_api.go. After that I saw a massive improvement, namely 100 txs per second.

@bcbrock, may you validate what I see by first commenting out the calls to crypto.CloseClient and then running the tests again? Please, put security.multithreading.enabled back to false. Thanks.

@muralisrini, we definitely need a way to address this issue related to crypto.CloseClient. Do you have already an idea on how to address this? We would need a thread that every once in a while checks when a client has been used last time and closes it if it hasn't been used for a certain period.

P.S.: @bcbrock, running your busywork framework (which is btw great) I see at some point the following lines:

driver ERROR Aborting on error described below:
driver ERROR Getting chaincode names failed for some reason; Aborting
make[1]: **\* [issue996b] Error 1

Do you know why? I see this error also when I put back crypto.CloseClient.

**Comment by bcbrock** [comment]
@adecaro I apologize for the delay in my reply - I have been traveling. I hacked the code to cause `crypto.CloseClient()` to simply return. The charts below show that this eliminates the "sawtooth" pattern and significantly improves transaction throughput. We still need to root-cause what makes certificate generation take so long. Sorry about the error above - that was a bug in my code that should be corrected now.
![chart1](https://cloud.githubusercontent.com/assets/11968347/14588229/854bd8a4-0489-11e6-8d4b-c4387b7561b4.png)
![chart2](https://cloud.githubusercontent.com/assets/11968347/14588231/8d898dae-0489-11e6-8aeb-6b18f3bc3c9b.png)

**Comment by adecaro** [comment]
@bcbrock, great job. @muralisrini, now we have to address in a systematic way the CloseClient issue.  
@diegomasini, do you have any news on the performance of generating TCerts? May we do it in parallel?

**Comment by adecaro** [comment]
Hi @bcbrock, I pushed a PR to address the CloseClient issue. May you test it?

**Comment by mastersingh24** [comment]
While I agree with the fact that we need to look at the generating / loading of TCerts, we need to be able to actually separate that from the fact that there is actually a different issue here and one that we should actually work out architecturally.

What you are really encountering here is a **client** transaction creating error not a **peer** processing issue, although due to the way the REST API works, they are current intertwined.

Assuming you are targeting the REST API of one of the validating peers, here's what is really happening under the covers at a high level.   Embedded within the peer code is actually a fabric client (this is part of the devops package).  If you run the peer in CLI mode and submit transactions, it is actually calling the devops package internally.  Now the REST API actually also happens to call the devops package underneath the covers.  So you have REST --> Devops Client --> peer going on.  Within the devops client, it call the crypto client which is responsible for obtaining the TCerts and also for creating / formatting proper fabric transactions.

So we need to think about a couple of things here:

1) What's the best way to actually host / run the REST API?  The original design intended for people to enable the REST API on a non-validating peer (which would still show the TCert generation issue) but would not bog down the actual peer

2) We actually need to have a native client SDK for being able to directly invoke transactions on peers.  The first example of this is underway with the JavaScript client SDK ( https://github.com/hyperledger/fabric/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen++node+client+sdk+ ) and in that case you can control how you deal with TCerts, batches, etc

3) A lighterweight REST "server" which stands in front of peers (basically a glorified client with a REST front end).  This would be similar to what is being done in terms of creating an online wallet API

4) We should be creating a native fabric client for performance testing.  With full control of the client API, you could for example invoke all transactions with the same TCert or download a bunch of TCerts and just cycle through them.  This way testing peer performance would not be dependent on client performance

**Comment by bcbrock** [comment]
@adecaro Did you ever resubmit the PR on the CloseClient issue?

**Comment by adecaro** [comment]
Hi @bcbrock, thanks for the reminder. Actually, I had submitted a PR for this but there were failures in travis and I withdrew it for further investigations. It is clearly time to resurrect the PR :) I will do asap

**Comment by bcbrock** [comment]
@adecaro Angelo, has the PR mentioned above been submitted/merged?

**Comment by christo4ferris** [comment]
@bcbrock @adecaro status on this one please? Please reference the PR if it has been submitted and update  (or include in) the PR description to indicate that it fixes this issue.

**Comment by bcbrock** [comment]
John Harrison of IBM ( @harrijk I believe ) has root-caused part of the problem as an inefficient algorithm for handling the TCert database - This problem generates the observed periodicity in throughput. I'm not aware of any fixes for this; Also don't know what happened to @adecaro fix for the CloseClient issue.

**Comment by bcbrock** [comment]
@christo4ferris With the transition to the new SDK this issue is no longer relevant or being worked by the security developers. Closing.
